concepts/iterators/introduction.md

```diff
@@ -0,0 +1,194 @@
+}
+```
+
+Many of Rust's [`collections`](https://doc.rust-lang.org/std/collections/index.html) implement
+[`IntoIterator`](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html),
+which can produce an `Iterator` from the collection's values.
+
+```rust
+trait IntoIterator where Self::IntoIter: Iterator<Item=Self::Item> {
+    type Item;
+    type IntoIter: Iterator;
+    fn into_iter(self) -> Self::IntoIter;
+}
+```
+
+Note that while `IntoIterator::into_iter` consumes its `self` type, that type does not necessarily have to be an owned value.
+Thus, there are `IntoIterator` implementations for [`Vec`](https://doc.rust-lang.org/std/vec/struct.Vec.html#impl-IntoIterator-2),
+for [`&'a Vec`](https://doc.rust-lang.org/std/vec/struct.Vec.html#impl-IntoIterator),
+and for [`&'a mut Vec`](https://doc.rust-lang.org/std/vec/struct.Vec.html#impl-IntoIterator-1).
+These implementations for varying degrees of ownership provide useful flexibility.
+
+An understanding of `traits` is not necessary for this concept and we will have a whole exercise for you to master `traits`.
+For now, just remember that:
+
+- Any type that implements `IntoIterator` is called an _iterable_ because it can be turned into an iterator.
+- Any type that implements `Iterator` is an iterator, which _produces_ values and can be _iterated over_.
+- You can get an iterator from any type that implements `IntoIterator` by calling its `into_iter` method.
+- Every iterator [implements `IntoIterator`](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html#impl-IntoIterator-25) and produces itself.
+- `for` loops always operate on iterators and call `IntoIterator` implicitly on their range argument.
+
+## `into_iter` and for loops
+
+You're probably familiar with for loops over a vector:
+
+```rust
+// lor is a `Vec`, which implements `IntoIterator`
+let lor = vec!["Ainur", "Elves", "Men", "Dwarves", "Hobbits", "Orcs", "Dragons"];
+for people in &lor {
+    println!("{people}");
+}
+```
+
+Fundamentally, a `for` loop is just shorthand for calling its argument's `into_iter` method,
+and then calling `next()` until it stops producing items. In other words, the previous loop desugars as:
+
+```rust
+{
+    let mut iterator = (&lor).into_iter();
+    while let Some(people) = iterator.next() {
+        println!("{people}");
+    }
+}
+
+## `iter` and `iter_mut` methods
+
+Most Rust collection types provide `iter` and `iter_mut` methods.
+For a collection of type `T`, these two methods perform the following by convention:
+
+- The iterator returned by `iter` will yield immutably borrowed `&T`.
+- The iterator returned by `iter_mut` will yield mutably borrowed `&mut T`.
+
+```rust
+let mut v = vec![1, 2, 3];
+
+// Using `iter` method
+let mut iterator = v.iter();
+assert_eq!(iterator.next(), Some(&1));
+assert_eq!(iterator.next(), Some(&2));
+assert_eq!(iterator.next(), Some(&3));
+assert_eq!(iterator.next(), None);
+
+// Using `iter_mut` method
+let mut muterator = v.iter_mut();
```

Comment:
It might be worth demoing why you'd ever use `iter_mut`; something like

```rust
for value in v.iter_mut() {
     *value *= *value;
}
println!("{v}");
// [1, 4, 9]
```